core/src/main/java/org/elasticsearch/search/internal/SearchContext.java

```diff
@@ -114,6 +120,18 @@ public ParseFieldMatcher parseFieldMatcher() {
         return parseFieldMatcher;
     }
 
+    public void setTask(SearchTask task) {
+        this.task = task;
+    }
+
+    public SearchTask getTask() {
+        return task;
+    }
+
+    public boolean isCancelled() {
+        return task != null && task.isCancelled();
```

Comment:
I don't think task should ever be null at that point? Should we just assert that `task` is not null?